stats.go

```diff
@@ -316,26 +323,28 @@ func (s *statStore) Flush() {
 
 		s.sink.FlushCounter(name, value)
 	}
+	s.countersMtx.RUnlock()
 
+	s.gaugesMtx.RLock()
 	for name, gv := range s.gauges {
 		value := gv.Value()
 		s.sink.FlushGauge(name, value)
 	}
+	s.gaugesMtx.RUnlock()
 }
 
 func (s *statStore) Start(ticker *time.Ticker) {
 	s.run(ticker)
 }
 
 func (s *statStore) AddStatGenerator(statGenerator StatGenerator) {
-	s.Lock()
-	defer s.Unlock()
-
+	s.genMtx.Lock()
 	s.statGenerators = append(s.statGenerators, statGenerator)
+	s.genMtx.Unlock()
```

Comment:
`defer` incurs a slight performance penalty. In this case, since it's such a short method with no return, there's no reason to use it.